code/contracts/ONEWallet2.sol

```diff
@@ -0,0 +1,157 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.8.0;
+
+contract ONEWallet2 {
+
+    bytes32 root; // Note: @ivan brought up a good point in reducing this to 16-bytes so hash of two consecutive nodes can be done in a single word (to save gas and reduce blockchain clutter). Let's not worry about that for now and re-evalaute this later.
+    uint8 height; // including the root. e.g. for a tree with 4 leaves, the height is 3.
+    uint8 interval; // otp interval in seconds, default is 30
+    uint32 t0; // starting time block (effectiveTime (in ms) / interval)
+    uint32 lifespan;  // in number of blocks (e.g. 1 block per [interval] seconds)
+    
+    address payable lastResortAddress; // where money will be sent during a recovery process (or when the wallet is beyond its lifespan)
+    bytes32 recoveryHash; // hash of the secret stored at the client, which can be used to recover the walletif the user loses the authenticator
+    uint8 constant MAX_REVEAL_DELAY = 2; // in OTP intervals (i.e., 2 * interval seconds)    
+    
+    // Management of daily spendings
+    uint256 dailyLimit; // uint128 is sufficient, but uint256 is more efficient since EVM works with 32-byte words.
+    mapping(uint => uint) dailySpendings; // day => amount ; can be cleaned up once in a while to save storage
+    uint cleanupCheckpoint; // the index of the day since epoch, when the last cleanup was made
+    uint16 constant CLEAN_UP_DAILY_SPENDING_BATCH = 1024; // this is to avoid out-of-gas exceptions
+    uint8 AUTO_CLEANUP_DAYS = 100; // do the auto cleanup of daily spending data after every 100 days
+    
+    uint lastExecutedIdx; // stores the index of OTP that was executed last (this is to avoid reply attacks within the same OTP interval by not allowing commits <= than this number)
+
+    enum OperType { TRANSFER, RECOVERY, SET_LAST_RESORT_ADDR }
+
+    struct Commit {
```

Comment:
I think we should hide the nature of the commit as much as possible. A user's intention should not be revealed in many use cases. For example, in a bidding or betting scenario, the time gap between an exposed commit and a reveal would put the user at some disadvantage.

Reply:
Resolved, since this seems to have been updated. Previously, with the parameters exposed, the issue is that an adversary knowing a user's intent (and exact amount) to bid, could carefully place a slightly higher bid between commit and reveal. A similar scenario can be formulated in adversarial games. The user who (is forced to) announces its intent earlier than others are put at a disadvantage.